### PR TITLE
Port `JsonCodec` from `circe-generic`

### PR DESCRIPTION
**Pull request by backuitist**

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -10,8 +10,9 @@ circe-generic in that this derivation is not based on Shapeless's generic repres
 classes and ADTs, and it is not intended to be as fully featured as `io.circe.generic`. In
 particular:
 
-* It's not published for Scala 2.10.
+* Sealed hierarchies (ADTs) aren't supported
```
Comment (travisbrown) [inline README.md]: For what it's worth this is coming soon (but I agree it's good to note it for now).

File: build.sbt [hunk 2]
```diff
@@ -77,7 +78,8 @@ lazy val derivationBase = crossModule("derivation", CrossType.Full)
       "io.circe" %%% "circe-testing" % circeVersion % Test
     ),
     ghpagesNoJekyll := true,
-    docMappingsApiDir := "api"
+    docMappingsApiDir := "api",
+    addCompilerPlugin("org.scalamacros" % "paradise" % paradiseVersion cross CrossVersion.patch)
```
Comment (travisbrown) [inline build.sbt]: Hmm, ugh, I forgot about the plugin. I think we should go ahead and merge this, but we might consider splitting it out into a separate module before 1.0.

If I remember correctly, though, there should be no problem depending on and using circe-derivation without enabling the plugin as long as you don't use the annotations? I should definitely confirm this.

File: derivation/shared/src/test/scala/io/circe/derivation/JsonCodecMacrosSuite.scala [hunk 3]
```diff
@@ -0,0 +1,159 @@
...
+          a <- A.arbitrary
+          b <- B.arbitrary
+          j <- Arbitrary.arbitrary[Int]
+        } yield Typed2(i, a, b, j)
+      )
+  }
+
+  // Hierarchy
+
+//  @JsonCodec sealed trait Hierarchy
```
Comment (travisbrown) [inline derivation/shared/src/test/scala/io/circe/derivation/JsonCodecMacrosSuite.scala]: Out of curiosity, why is this commented out?

File: build.sbt (hunk 2 above)
Comment (backuitist) [reply build.sbt]: Yes, very much like `circe-generic` the plugin should only be needed if you use of the annotation. That should be documented though.

File: derivation/shared/src/test/scala/io/circe/derivation/JsonCodecMacrosSuite.scala (hunk 3 above)
Comment (travisbrown) [reply derivation/shared/src/test/scala/io/circe/derivation/JsonCodecMacrosSuite.scala]: Oh, of course—reading too quickly this morning.

Comment (backuitist) [reply derivation/shared/src/test/scala/io/circe/derivation/JsonCodecMacrosSuite.scala]: No worries :) Out of curiosity, how do you plan on adding the support for ADTs?

**Pull request merged**